datalad_next/url_operations/ssh.py

```diff
@@ -81,63 +90,63 @@ def stat(self,
         See :meth:`datalad_next.url_operations.UrlOperations.stat`
         for parameter documentation and exception behavior.
         """
+        ssh_cat = _SshCommandBuilder(url)
+        cmd = ssh_cat.get_cmd(SshUrlOperations._stat_cmd)
         try:
-            props = self._stat(
-                url,
-                cmd=SshUrlOperations._stat_cmd,
-            )
-        except CommandError as e:
-            if e.code == 244:
-                # this is the special code for a file-not-found
-                raise UrlOperationsResourceUnknown(url) from e
-            else:
-                raise UrlOperationsRemoteError(url, message=str(e)) from e
-
+            with iter_subproc(cmd) as stream:
+                # any exception that is raised in this context and not caught
+                # will prevent the creation of `IterableSubprocessError`. But
+                # we rely on the return code of the ssh-process to signal
+                # specific errors. Therefore, we catch the expected
+                # `StopIteration` here.
+                try:
+                    props = self._get_props(url, stream)
+                except StopIteration:
+                    pass
```

Comment:
An example without an exception in the `iter_subproc`-context:

```python
>>> from datalad_next.runners.iter_subproc import iter_subproc
>>> with iter_subproc(['ls', '/THIS_DOES_NOT_EXIST']) as sub:
...     tuple(sub)
... 
()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib64/python3.10/contextlib.py", line 142, in __exit__
    next(self.gen)
  File "/home/cristian/Develop/datalad-next/datalad_next/iterable_subprocess/iterable_subprocess.py", line 153, in iterable_subprocess
    raise IterableSubprocessError(proc.returncode, b''.join(stderr_deque)[-chunk_size:])
datalad_next.iterable_subprocess.iterable_subprocess.IterableSubprocessError: (2, b"ls: Zugriff auf '/THIS_DOES_NOT_EXIST' nicht m\xc3\xb6glich: Datei oder Verzeichnis nicht gefunden\n")
```

This raises `IterableSubprocessError`.

The following example raises `StopIteration`:

```python
>>> from datalad_next.runners.iter_subproc import iter_subproc
>>> with iter_subproc(['ls', '/THIS_DOES_NOT_EXIST']) as sub:
...     next(sub)
...
Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
StopIteration
```

Because we want to get the return code of our subprocesses,  we need to catch `StopIteration` (in this case it is `StopIteration`, but generally all exceptions that are raised in the `iter_subproc`-context need to be caught to get to an `IterableSubprocessError`).

We could of course write code that does not raise `StopIteration`, but that requires to pass error values around, which is not too nice either.

Reply:
But how do the examples relate to the usage pattern in this PR? The iteration is done within the for-loop. How would `StopIteration` get out? My last example raises inside the for-loop, and we see `RuntimeError`, the actual exception.

If I understand you correctly, `IterableSubprocessError` would never come out, unless all other exceptions are caught. But I do not see that (and it would be a bit weird):

```py
In [1]: from datalad_next.runners import iter_subproc
In [2]: with iter_subproc(['ls', '/THIS_DOES_NOT_EXIST']) as sub:
   ...:     for i in sub:
   ...:         print('some')
   ...: 
---------------------------------------------------------------------------
IterableSubprocessError                   Traceback (most recent call last)
Cell In[2], line 1
----> 1 with iter_subproc(['ls', '/THIS_DOES_NOT_EXIST']) as sub:
      2     for i in sub:
      3         print('some')

File /usr/lib/python3.11/contextlib.py:144, in _GeneratorContextManager.__exit__(self, typ, value, traceback)
    142 if typ is None:
    143     try:
--> 144         next(self.gen)
    145     except StopIteration:
    146         return False

File ~/hacking/datalad/next/datalad_next/iterable_subprocess/iterable_subprocess.py:153, in iterable_subprocess(program, input_chunks, chunk_size)
    150         raise e.__context__ from None
    152 if proc.returncode:
--> 153     raise IterableSubprocessError(proc.returncode, b''.join(stderr_deque)[-chunk_size:])

IterableSubprocessError: (2, b"ls: cannot access '/THIS_DOES_NOT_EXIST': No such file or directory\n")
```

Reply:
> But how do the examples relate to the usage pattern in this PR? The iteration is done within the for-loop. How would `StopIteration` get out? My last example raises inside the for-loop, and we see `RuntimeError`, the actual exception.
> [...]

We also use `next()` in `self._get_props()`. We can use a for-loop for those as well, but then we have to pass an error-result, e.g. `None` to the caller and handle that in the the caller.

> If I understand you correctly, `IterableSubprocessError` would never come out, unless all other exceptions are caught. [,,,]

Yes! `IterableSubprocessError` will never come out, if any exception is raised in the `iter_subproc`-context (there is one exception from that rule, the internal `_BrokenPipeError`). E.g:

```python
>>> from datalad_next.runners.iter_subproc import iter_subproc
>>> with iter_subproc(['ls', '/THIS_DOES_NOT_EXIST']) as sub:
...     raise(ValueError)
... 
Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
ValueError
```
So we have to prevent exception-raising, if we want to get access to the return-code of the subprocess.
 

> [...] But I do not see that (and it would be a bit weird):
> 
> ```python
> In [1]: from datalad_next.runners import iter_subproc
> In [2]: with iter_subproc(['ls', '/THIS_DOES_NOT_EXIST']) as sub:
>    ...:     for i in sub:
>    ...:         print('some')
>    ...: 
> ---------------------------------------------------------------------------
> IterableSubprocessError                   Traceback (most recent call last)
> Cell In[2], line 1
> ----> 1 with iter_subproc(['ls', '/THIS_DOES_NOT_EXIST']) as sub:
>       2     for i in sub:
>       3         print('some')
> 
> File /usr/lib/python3.11/contextlib.py:144, in _GeneratorContextManager.__exit__(self, typ, value, traceback)
>     142 if typ is None:
>     143     try:
> --> 144         next(self.gen)
>     145     except StopIteration:
>     146         return False
> 
> File ~/hacking/datalad/next/datalad_next/iterable_subprocess/iterable_subprocess.py:153, in iterable_subprocess(program, input_chunks, chunk_size)
>     150         raise e.__context__ from None
>     152 if proc.returncode:
> --> 153     raise IterableSubprocessError(proc.returncode, b''.join(stderr_deque)[-chunk_size:])
> 
> IterableSubprocessError: (2, b"ls: cannot access '/THIS_DOES_NOT_EXIST': No such file or directory\n")
> ```
> 
> Take a look at the traceback ^^^

IIANM, the traceback does not involve an exception raised in the context (the value of `typ` in `_GeneratorContextManager.__exit__` is `None`), therefore lines 152 - 153 are executed.

Any exception in the `with iter_subproc`-context would trigger execution of lines 138 - 139 of `iterable_subprocess.py`:
```python
   137              except BaseException:
   138                  proc.terminate()
   139                  raise
```
And lines 152 - 153 are never executed

Reply:
OK, I think I am slowly getting what you are saying. I believe the key is:

> We also use next() in self._get_props().

But to me all the wrapping in try/except seems to be *solely*, because that is the case, and not "also".

If that is true, it would seem to be more sensible to prevent `_get_props()` from ever raising `StopIteration`. It is an internal function that always needs this protection, and that need is neither obvious, nor documented.

But that seems to require a rather complex logic, because, again if I understand this right, we would want to leave the `iter_subproc` context immediately, and have an outer shell catch the then generated `IterableSubprocessError` -- which is only generated, if there is no exception already.

If that is all correct, we need to discuss two things:

- this should be documented in `iter_subproc()` -- it is all but obvious
- can this be simplified (avoid having to wrap any and all context body like this) by also handling `StopIteration` where `_BrokenPipeError` is handled in `iterable_subprocess()`? Basically:

```py

    except _BrokenPipeError as e:
        if proc.returncode == 0:
            raise e.__context__ from None
    # add this ----------------
    except StopIteration:
       # something try to read output, but there is none
       # if that is a problem the next check will tell
    # until here ---------------

    if proc.returncode:
        raise IterableSubprocessError(proc.returncode, b''.join(stderr_deque)[-chunk_size:])
```

Reply:
I think that is a good approach (dedicated PR first, and this one waiting).

I will keep trying to think of a reason why we would not want to change `iterable_subprocesses`. Obviously, fewer changes are better, but this issues seems to be impossible to fix from the outside.

Conceptually, `_BrokenPipeError` and `StopIteration` appear to be to variants of the same problem: premature death. It continues to appear sensible to me to also deal with them in the same way.

You patch is a minimal as it can get. What I cannot anticipate are sideeffect of raising the `StopIteration.__context__`. Never seen that being done.

Reply:
> [...] You[r] patch is a minimal as it can get. What I cannot anticipate are sideeffect of raising the StopIteration.__context__. Never seen that being done.

That is a good point because we should not do that. So the patch will have to be a little more intrusive, something like:
```
    except _BrokenPipeError as e:
        if proc.returncode == 0:
            raise e.__context__ from None
    except StopIteration:
        if proc.returncode == 0:
            raise
```

> [...]
> Conceptually, `_BrokenPipeError` and `StopIteration` appear to be to variants of the same problem: premature death. It continues to appear sensible to me to also deal with them in the same way.
> [...]

That is a nice thought. I was struggeling to find sound reasoning for the patch.

Reply:
> [...] 
> Preparing the PR now.

I came across another idea on how to make return codes accessible in the presence of exceptions: PR #565.

(I had this idea after opening PR #562, which implements what is outlined above. So I put PR #562 into draft-mode)